geomstats/special_orthogonal_group.py

```diff
@@ -557,94 +557,352 @@ def matrix_from_quaternion(self, quaternion):
+                           and order == 'xyz'))
+
+        condition_2 = ((extrinsic_or_intrinsic == 'extrinsic'
+                        and order == 'xyz')
+                       or (extrinsic_or_intrinsic == 'intrinsic'
+                           and order == 'zyx'))
+        if condition_1:
+            rot_mat = self.matrix_from_tait_bryan_angles_extrinsic_zyx(
+                tait_bryan_angles)
+        elif condition_2:
+            rot_mat = self.matrix_from_tait_bryan_angles_extrinsic_xyz(
+                tait_bryan_angles)
 
-        yaw = yaw_pitch_roll[:, 0]
-        pitch = yaw_pitch_roll[:, 1]
-        roll = yaw_pitch_roll[:, 2]
+        else:
+            raise ValueError('extrinsic_or_intrinsic should be'
+                             ' \'extrinsic\' or \'intrinsic\''
+                             ' and order should be \'xyz\' or \'zyx\'.')
 
-        cos_half_yaw = gs.cos(yaw / 2.)
-        sin_half_yaw = gs.sin(yaw / 2.)
-        cos_half_pitch = gs.cos(pitch / 2.)
-        sin_half_pitch = gs.sin(pitch / 2.)
-        cos_half_roll = gs.cos(roll / 2.)
-        sin_half_roll = gs.sin(roll / 2.)
+        return rot_mat
 
-        quaternion = gs.zeros((n_yaw_pitch_rolls, 4))
+    def tait_bryan_angles_from_matrix(self, rot_mat,
+                                      extrinsic_or_intrinsic='extrinsic',
+                                      order='zyx'):
+        """
+        Convert a rotation matrix rot_mat into the tait bryan angles,
+        [angle_1, angle_2, angle_3] in extrinsic (fixed) coordinate frame,
+        for the order zyx, i.e.:
+        rot_mat = X(angle_1).Y(angle_2).Z(angle_3)
+        where:
+        - X(angle_1) is a rotation of angle angle_1 around axis x.
+        - Y(angle_2) is a rotation of angle angle_2 around axis y.
+        - Z(angle_3) is a rotation of angle angle_3 around axis z.
+        """
+        rot_mat = gs.to_ndarray(rot_mat, to_ndim=3)
+        quaternion = self.quaternion_from_matrix(rot_mat)
+        tait_bryan_angles = self.tait_bryan_angles_from_quaternion(
+            quaternion,
+            extrinsic_or_intrinsic=extrinsic_or_intrinsic,
+            order=order)
 
-        cos_half_angle = (cos_half_yaw * cos_half_pitch * cos_half_roll
-                          + sin_half_yaw * sin_half_pitch * sin_half_roll)
+        return tait_bryan_angles
 
-        quaternion[:, 0] = cos_half_angle
+    def quaternion_from_tait_bryan_angles_intrinsic_xyz(
+            self, tait_bryan_angles):
+        assert self.n == 3, ('The quaternion representation'
+                             ' and the Tait-Bryan angles representation'
+                             ' do not exist'
+                             ' for rotations in %d dimensions.' % self.n)
+        tait_bryan_angles = gs.to_ndarray(tait_bryan_angles, to_ndim=2)
+        n_tait_bryan_angles, _ = tait_bryan_angles.shape
+        quaternion = gs.zeros((n_tait_bryan_angles, 4))
+
+        angle_1 = tait_bryan_angles[:, 0]
```

Comment:
nit: is there a way to make this maybe nicer by keeping the angle array and using linear algebra in the subsequent operations?